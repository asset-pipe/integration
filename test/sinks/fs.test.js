'use strict';

const express = require('express');
const { resolve } = require('path');
const supertest = require('supertest');
const Client = require('@asset-pipe/client');
const AssetServer = require('@asset-pipe/server');
const AssetSinkFS = require('@asset-pipe/sink-fs');
const {
    endWorkers,
} = require('../../node_modules/@asset-pipe/server/lib/utils');

async function startTestServer(sink) {
    const app = express();
    const assets = new AssetServer(sink);
    app.use(assets.router());
    return new Promise(resolve => {
        const server = app.listen(() => {
            resolve({
                server,
                port: server.address().port,
            });
        });
    });
}

function closeTestServer(server) {
    return new Promise(resolve => {
        server.once('close', () => resolve());
        server.kill();
    });
}

function clean(data) {
    const reg1 = /http:\/\/127\.0\.0\.1:[0-9]+\//g;
    const reg2 = /"(\/[A-Za-z]+)+\/asset-pipe/gm;
    const str = JSON.stringify(data, null, 2).replace(reg1, '/');
    return JSON.parse(str.replace(reg2, '"asset-pipe'));
}

afterAll(() => endWorkers());

describe('asset-pipe-sink-fs', () => {
    let server;
    let port;
    let client;
    let get;

    beforeAll(async () => {
        const path = resolve(__dirname, '../../assets/fs-sink');
        const sink = new AssetSinkFS({ path });
        const started = await startTestServer(sink);
        server = started.server;
        port = started.port;
        get = supertest(`http://127.0.0.1:${port}`).get;
        client = new Client({
            buildServerUri: `http://127.0.0.1:${port}`,
        });
    });

    afterAll(async () => {
        await closeTestServer(server);
    });

    test('javascript upload', async () => {
        expect.assertions(1);
        const upload = await client.uploadFeed([resolve('assets/a.js')]);

        expect(clean(upload)).toMatchSnapshot();
    });

    test('javascript upload feed fetching', async () => {
        expect.assertions(1);
        const upload = await client.uploadFeed([resolve('assets/a.js')]);
        const { body } = await get(`/feed/${upload.file}`);
        expect(clean(body)).toMatchSnapshot();
    });

    test('javascript bundling', async () => {
        expect.assertions(1);
        const upload = await client.uploadFeed([resolve('assets/a.js')]);
        const bundle = await client.createRemoteBundle([upload.file], 'js');

        expect(clean(bundle)).toMatchSnapshot();
    });

    test('javascript bundle fetching', async () => {
        expect.assertions(1);
        const upload = await client.uploadFeed([resolve('assets/a.js')]);
        const bundle = await client.createRemoteBundle([upload.file], 'js');
        const { body } = await get(`/feed/${bundle.file}`);

        expect(clean(body)).toMatchSnapshot();
    });

    test('css upload', async () => {
        expect.assertions(1);
        const feeds = [resolve('assets/style.css')];
        const upload = await client.uploadFeed(feeds);

        expect(clean(upload)).toMatchSnapshot();
    });

    test('css upload feed fetching', async () => {
        expect.assertions(1);
        const feeds = [resolve('assets/style.css')];
        const upload = await client.uploadFeed(feeds);
        const { body } = await get(`/feed/${upload.file}`);

        expect(clean(body)).toMatchSnapshot();
    });

    test('css bundling', async () => {
        expect.assertions(1);
        const feeds = [resolve('assets/style.css')];
        const upload = await client.uploadFeed(feeds);
        const bundle = await client.createRemoteBundle([upload.file], 'css');

        expect(clean(bundle)).toMatchSnapshot();
    });

    test('css bundle fetching', async () => {
        expect.assertions(1);
        const feeds = [resolve('assets/style.css')];
        const upload = await client.uploadFeed(feeds);
        const bundle = await client.createRemoteBundle([upload.file], 'css');
        const { body } = await get(`/feed/${bundle.file}`);

        expect(clean(body)).toMatchSnapshot();
    });
});
