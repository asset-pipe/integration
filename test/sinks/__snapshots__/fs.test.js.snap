// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asset-pipe-sink-fs css bundle fetching 1`] = `Object {}`;

exports[`asset-pipe-sink-fs css bundling 1`] = `
Object {
  "file": "8b30de6afdb4474cfd5c5df052408dc6ee170d243861c3fa63d246440b5f472d.css",
  "uri": "/bundle/8b30de6afdb4474cfd5c5df052408dc6ee170d243861c3fa63d246440b5f472d.css",
}
`;

exports[`asset-pipe-sink-fs css upload 1`] = `
Object {
  "file": "80c8a092aee329dd8cf852eaa081829bcac6f24bfa4ef11b90c94b9607244251.json",
  "id": "80c8a092aee329dd8cf852eaa081829bcac6f24bfa4ef11b90c94b9607244251",
  "uri": "/feed/80c8a092aee329dd8cf852eaa081829bcac6f24bfa4ef11b90c94b9607244251.json",
}
`;

exports[`asset-pipe-sink-fs css upload feed fetching 1`] = `
Array [
  Object {
    "content": "html, body, div, span, object, iframe,
h1, h2, h3, h4, h5, h6, p, blockquote, pre,
abbr, address, cite, code,
del, dfn, em, img, ins, kbd, q, samp,
small, strong, sub, sup, var,
b, i,
dl, dt, dd, ol, ul, li,
fieldset, form, label, legend,
table, caption, tbody, tfoot, thead, tr, th, td,
article, aside, canvas, details, figcaption, figure, 
footer, header, hgroup, menu, nav, section, summary,
time, mark, audio, video {
    margin:0;
    padding:0;
    border:0;
    outline:0;
    font-size:100%;
    vertical-align:baseline;
    background:transparent;
}

body {
    line-height:1;
}

article,aside,details,figcaption,figure,
footer,header,hgroup,menu,nav,section { 
    display:block;
}

nav ul {
    list-style:none;
}

blockquote, q {
    quotes:none;
}

blockquote:before, blockquote:after,
q:before, q:after {
    content:'';
    content:none;
}

a {
    margin:0;
    padding:0;
    font-size:100%;
    vertical-align:baseline;
    background:transparent;
}

/* change colours to suit your needs */

ins {
    background-color:#ff9;
    color:#000;
    text-decoration:none;
}

/* change colours to suit your needs */

mark {
    background-color:#ff9;
    color:#000; 
    font-style:italic;
    font-weight:bold;
}

del {
    text-decoration: line-through;
}

abbr[title], dfn[title] {
    border-bottom:1px dotted;
    cursor:help;
}

table {
    border-collapse:collapse;
    border-spacing:0;
}

/* change border colour to suit your needs */

hr {
    display:block;
    height:1px;
    border:0;   
    border-top:1px solid #cccccc;
    margin:1em 0;
    padding:0;
}

input, select {
    vertical-align:middle;
}

h1 {
    color: #000000;
    size: 32px;
}",
    "file": "integration/assets/style.css",
    "id": "8b30de6afdb4474cfd5c5df052408dc6ee170d243861c3fa63d246440b5f472d",
    "name": "integration",
    "version": "1.0.0",
  },
]
`;

exports[`asset-pipe-sink-fs javascript bundle fetching 1`] = `Object {}`;

exports[`asset-pipe-sink-fs javascript bundling 1`] = `
Object {
  "file": "2cfa0e02247eaf478f30078a443e90e83a7fe1b56489e63fb5f7a6be6f595168.js",
  "uri": "/bundle/2cfa0e02247eaf478f30078a443e90e83a7fe1b56489e63fb5f7a6be6f595168.js",
}
`;

exports[`asset-pipe-sink-fs javascript upload 1`] = `
Object {
  "file": "149967898754950cb6001298f67fbaa674420d8aa409c48a7ef2c5fb675843d0.json",
  "id": "149967898754950cb6001298f67fbaa674420d8aa409c48a7ef2c5fb675843d0",
  "uri": "/feed/149967898754950cb6001298f67fbaa674420d8aa409c48a7ef2c5fb675843d0.json",
}
`;

exports[`asset-pipe-sink-fs javascript upload feed fetching 1`] = `
Array [
  Object {
    "deps": Object {
      "./d": "320f596434a079912e951a037302087837f241f2896be048e3ccd18d8e6888c9",
    },
    "entry": true,
    "expose": false,
    "file": "asset-pipe/integration/assets/a.js",
    "id": "893576861db8f16e03815f618ca6838cb4af6fd867b10ffa5684e1f1fcd6b467",
    "index": 1,
    "indexDeps": Object {
      "./d": 2,
    },
    "order": 0,
    "source": "'use strict';

const sum = require('./d');

sum(1, 2);
",
  },
  Object {
    "deps": Object {},
    "file": "asset-pipe/integration/assets/d.js",
    "id": "320f596434a079912e951a037302087837f241f2896be048e3ccd18d8e6888c9",
    "index": 2,
    "indexDeps": Object {},
    "source": "'use strict';

module.exports = function sum(a, b) {
    return a + b;
};
",
  },
]
`;
